lib/customer-deposit-wallet/src/Cardano/Wallet/Deposit/Pure.hs

```diff
@@ -345,23 +366,103 @@ wonders interval =
     Operations
-    Writing to blockchain
+    Constructing transactions
 ------------------------------------------------------------------------------}
+data ErrCreatePayment
+    = ErrCreatePaymentNotRecentEra (Read.EraValue Read.Era)
+    | ErrCreatePaymentBalanceTx (Write.ErrBalanceTx Write.Conway)
+    deriving (Eq, Show)
 
+-- | Create a payment to a list of destinations.
 createPayment
-    :: [(Address, Write.Value)] -> WalletState -> Maybe Write.TxBody
-createPayment = undefined
+    :: Read.EraValue Read.PParams
+    -> Write.TimeTranslation
+    -> [(Address, Write.Value)]
+    -> WalletState
+    -> Either ErrCreatePayment Write.Tx
+createPayment (Read.EraValue (Read.PParams pparams :: Read.PParams era)) a b w =
+    case Read.theEra :: Read.Era era of
+        Read.Conway ->
+            first ErrCreatePaymentBalanceTx
+                $ createPaymentConway pparams a b w
+        era' -> Left $ ErrCreatePaymentNotRecentEra (Read.EraValue era')
+
+-- | In the Conway era: Create a payment to a list of destinations.
+createPaymentConway
+    :: Write.PParams Write.Conway
+    -> Write.TimeTranslation
+    -> [(Address, Write.Value)]
+    -> WalletState
+    -> Either (Write.ErrBalanceTx Write.Conway) Write.Tx
+createPaymentConway pparams timeTranslation destinations w =
+    fmap (Read.Tx . fst)
+    . flip Random.evalRand (pilferRandomGen w)
+    . runExceptT
+    . balance
+        (UTxOHistory.getUTxO $ utxoHistory w)
+        (addresses w)
+    . mkPartialTx
+    $ paymentTxBody
+  where
+    paymentTxBody :: Write.TxBody
+    paymentTxBody = Write.TxBody
+        { spendInputs = mempty
+        , collInputs = mempty
+        , txouts =
+            Map.fromList $ zip [(toEnum 0)..] $ map (uncurry Write.mkTxOut) destinations
+        , collRet = Nothing
+        }
+
+    mkPartialTx :: Write.TxBody -> Write.PartialTx Write.Conway
+    mkPartialTx txbody = Write.PartialTx
+        { tx = Read.unTx $ Write.mkTx txbody
+        , extraUTxO = mempty :: Write.UTxO Write.Conway
+        , redeemers = mempty
+        , stakeKeyDeposits = Write.StakeKeyDepositMap mempty
+        , timelockKeyWitnessCounts = Write.TimelockKeyWitnessCounts mempty
+        }
+
+    balance utxo addressState =
+        Write.balanceTx
+            pparams
+            timeTranslation
+            Write.AllKeyPaymentCredentials
+            (Write.constructUTxOIndex $ Write.toConwayUTxO utxo)
+            (changeAddressGen addressState)
+            ()
+
+    changeAddressGen s = Write.ChangeAddressGen
+        { Write.genChangeAddress =
+            first Read.decompactAddr . Address.newChangeAddress s
+        , Write.maxLengthChangeAddress =
+            Read.decompactAddr $ Address.mockMaxLengthChangeAddress s
+        }
+
+-- | Use entropy contained in the current 'WalletState'
+-- to construct a pseudorandom seed.
+-- (NOT a viable source of cryptographic randomness.)
+pilferRandomGen :: WalletState -> Random.StdGen
```

Comment:
🤔 My thinking was that `balanceTx` uses a randomized algorithm to improve the result, but the presence or absence of a result should not materially depend on randomization.

1) Hm. In principle, it could be possible to reverse-engineer information about the `StdGen` and therefore information about the `ChainPoint` from the observed selection of coins.
2) You're probably referring to the fact that concurrent calls to `balanceTx` with the same `UTxO` may have a higher likelihood of selecting different inputs.
3) Similar to 2). The necessity to retry would indicate a material defect in the `balanceTx`, though. ("Sometimes it gives a result, sometimes it doesn't".)

Alternative designs:

* Keep track of a `StdGen` in `WalletState`. Avoids 1.
* Use a cryptographic hash of the transaction as source of randomness. Avoids 1 and 2.

The point 3 cannot be changed in a stateless manner — you'd need to know about previous attempts at calling `balanceTx`. A fast-evolving source of randomness can mitigate this, but I think that it's better to have `balanceTx` not fail randomly.

Reply:
> 1. Hm. In principle, it could be possible to reverse-engineer information about the StdGen and therefore information about the ChainPoint from the observed selection of coins.

I was more thinking that: if `utxo, walletTip, guessPartialTx onchainTx` can determinsitically produce the resulting `onchainTx`, it could make correlating enterprises from the same wallet easer, and possibly in some new sense leak information about the utxo set as a whole.  However... one would have to guess the `walletTip` at the time of tx construction, which might make the approach worthless even with just a few likely tips.

> You're probably referring to the fact that concurrent calls to balanceTx with the same UTxO may have a higher likelihood of selecting different inputs.

*the same inputs, yes. And I think there's a decent chance it's guaranteed. But... this may possibly not be terrible as you shouldn't rely on randomness to prevent selecting the same inputs anyway 🤷‍♂️ 

> Similar to 2). The necessity to retry would indicate a material defect in the balanceTx, though. ("Sometimes it gives a result, sometimes it doesn't".)

Not _necessarily_. There could be cases where the user as strong opinions on the selection and would like to discard it to try again. 

As for retries not being an intended solution for failing balanceTx results, I do agree. (Whether `balanceTx` should be _guaranteed_ to find the only succeeding selection in tricky situations if such exists is another question though. I suspect we only need/want some `balance >= paymentAmt + delta => can make payment` property to hold, and are ok with the result sometimes succeeding and sometimes failing when `paymentAmt <= balance <= paymentAmt + delta`)

> Keep track of a StdGen in WalletState. Avoids 1 by revealing only uninteresting information.

If you provide a non-public initial seed, probably 🤔 

> Use a cryptographic hash of the transaction as source of randomness. Avoids 1 and 2.

What tx? The one that isn't constructed yet? Identical _partial_ txs may be balanced and submitted repeatedly. We could use all arguments to `balanceTx` as source of randomness though (in particular the utxo set). This wouldn't avoid 1 though (or it might depend on how far-fetched or realistic 1 is).

Reply:
Anyway, happy to go with the approach you have, as it's interesting and I see no clear problem with it, but I did want to raise the points.